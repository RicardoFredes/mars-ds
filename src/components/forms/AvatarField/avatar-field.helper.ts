export const dictionary = {
  alt: "Foto de perfil",
  overlayText: "Adicionar foto",
  acceptMimeTypes: "image/*",
  caption: (extensions: string[], maxSize: number): string => {
    const ext = extensions.map((extension) => `*.${extension}`).join(", ");
    const sizeInMB = Math.round(maxSize / 1024 / 1024);

    return `Arquivos ${ext}<br/>Tamanho m√°ximo ${sizeInMB} MB`;
  },
};
