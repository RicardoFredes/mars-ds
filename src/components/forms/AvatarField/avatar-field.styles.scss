.avatar-field {
  display: flex;
  flex-direction: column;
  gap: $gap-sm-4;
  align-items: center;
  justify-content: center;
  color: $color-text-secondary;

  &__content {
    display: grid;
    grid-template-rows: 1fr;
    grid-template-columns: 1fr;
    align-items: center;
    width: 144px;
    height: 144px;
    padding: $spacing-sm;
    cursor: pointer;
    border: 1px dashed $color-divider;
    border-radius: 50%;
  }

  &__overlay {
    display: flex;
    flex-direction: column;
    gap: $gap-sm-2;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    padding: $spacing-md;
    border-radius: 50%;

    &:hover {
      background-color: $color-neutral-100;
      transition: all 0.25s ease;
    }

    &--has-image {
      color: transparent;
      background-color: transparent;

      &:hover {
        color: $color-white;
        background-color: $color-background-overlay;
      }
    }
  }

  &__input {
    display: none;
  }

  &__preview {
    width: 100%;
    height: 100%;
    overflow: hidden;
    border-radius: 50%;
    object-fit: cover;
  }

  &__input,
  &__overlay,
  &__preview {
    grid-row: 1;
    grid-column: 1;
  }

  &__caption {
    text-align: center;
  }
}

.theme-dark {
  .avatar-field {
    &__overlay {
      &:hover {
        color: $color-white;
        background-color: $color-neutral-300;
        transition: all 0.25s ease;
      }

      &--has-image {
        &:hover {
          background-color: $color-background-overlay;
        }
      }
    }
  }
}
