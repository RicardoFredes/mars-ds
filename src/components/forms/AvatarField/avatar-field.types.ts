import type { FileInputProps as CommonFileFieldProps } from "@/components/primitives/FileInput";

interface AvatarFieldProps extends CommonFileFieldProps {
  extensions: string[];
}

export type { AvatarFieldProps };
