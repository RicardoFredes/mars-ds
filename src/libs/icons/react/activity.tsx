/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconActivity = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  strokeLinejoin="round"
>
  <polyline points="22 12 18 12 15 21 9 3 6 12 2 12" />
</svg>
);
  
export default IconActivity;