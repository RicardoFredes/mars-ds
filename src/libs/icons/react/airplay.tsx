/**
 * WARNING!
 * Do NOT edit this file as it is automatically generated
 * run "yarn generate:icons" to update
 */

  
import type { IconLibProps } from "@/libs/icons/icons.types";

const IconAirplay = ({ height = 24, width = 24 }: IconLibProps) => (
<svg
  xmlns="http://www.w3.org/2000/svg"
  width={width}
  height={height}
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  strokeWidth="2"
  strokeLinecap="round"
  stroke-linejoin="round"
>
  <path d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1" />
  <polygon points="12 15 17 21 7 21 12 15" />
</svg>
);
  
export default IconAirplay;