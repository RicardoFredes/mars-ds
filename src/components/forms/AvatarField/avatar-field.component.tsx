import type { AvatarFieldProps } from "./avatar-field.types";

import classNames from "classnames";
import { useState } from "react";

import Icon from "@/components/basics/Icon";
import FileInput, { FIVE_MB } from "@/components/primitives/FileInput";
import Caption from "@/components/typographies/Caption";
import Text, { TextSizes } from "@/components/typographies/Text";

import { dictionary } from "./avatar-field.helper";

const AvatarField = ({
  className,
  extensions = ["jpg, jpeg, png"],
  maxSize = FIVE_MB,
  src,
  ...props
}: AvatarFieldProps) => {
  const [file, setFile] = useState<File | null>(null);

  const cn = classNames("avatar-field", className);
  const overlayCn = classNames("avatar-field__overlay", {
    "avatar-field__overlay--has-image": file,
  });

  const captionText = dictionary.caption(extensions, maxSize);

  const handleSelectFile = (file: File | null) => setFile(file);

  return (
    <div className={cn} {...props}>
      <label className="avatar-field__content">
        <FileInput
          name="file"
          title={dictionary.alt}
          accept={dictionary.acceptMimeTypes}
          className="avatar-field__input"
          onSelectFile={handleSelectFile}
          maxSize={maxSize}
          {...props}
        />

        {(file || src) && (
          <img
            src={file ? URL.createObjectURL?.(file) : src}
            alt={dictionary.alt}
            className="avatar-field__preview"
          />
        )}

        <div className={overlayCn}>
          <Icon name="camera" />
          <Text size={TextSizes.Small}>{dictionary.overlayText}</Text>
        </div>
      </label>

      <Caption
        className="avatar-field__caption"
        data-testid="avatar-field-caption"
        dangerouslySetInnerHTML={{ __html: captionText }}
      />
    </div>
  );
};

export default AvatarField;
