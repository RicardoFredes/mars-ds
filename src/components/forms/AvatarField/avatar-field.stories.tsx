import type { ComponentMeta, ComponentStory } from "@storybook/react";

import { dictionary } from "@/components/primitives/FileInput";
import { commonStoreComponentProps } from "@/services/stories";

import AvatarField from "./avatar-field.component";
export default {
  title: "Forms/AvatarField",
  component: AvatarField,
  argTypes: {
    extensions: {
      description: "The allowed file extensions",
      control: "array",
    },
    maxSize: {
      description: "The max file size",
      type: { name: "number", required: false },
    },
    notifications: {
      description: "Notifications to be displayed on certain events",
      control: "object",
      defaultValue: {
        ...dictionary.notifications,
      },
    },
    ...commonStoreComponentProps,
  },
} as ComponentMeta<typeof AvatarField>;

const Template: ComponentStory<typeof AvatarField> = (args) => <AvatarField {...args} />;

const DefaultArgs = {
  extensions: ["jpg", "jpeg", "png"],
};

export const Default = Template.bind({});
Default.args = DefaultArgs;

export const CustomMaxFileSize = Template.bind({});
CustomMaxFileSize.args = { ...DefaultArgs, maxSize: 10 * 1024 * 1024 /* 10MB */ };
CustomMaxFileSize.story = {
  parameters: {
    loki: { skip: true },
  },
};

export const CustomNotifications = Template.bind({});
CustomNotifications.args = {
  ...DefaultArgs,
  maxSize: 1 * 1024 * 1024,
  notifications: {
    tooBig: "O arquivo Ã© muito grande",
  },
};
CustomNotifications.story = {
  parameters: {
    loki: { skip: true },
  },
};
