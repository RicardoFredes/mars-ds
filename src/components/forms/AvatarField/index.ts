/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 * run "yarn reindex" to update
 */

import AvatarField from "./avatar-field.component";
export default AvatarField;
export * from "./avatar-field.types";
export * from "./avatar-field.helper";
