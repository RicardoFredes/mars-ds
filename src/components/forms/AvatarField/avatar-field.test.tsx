import type { AvatarFieldProps } from "./avatar-field.types";

import { render } from "@testing-library/react";

import AvatarField from "./avatar-field.component";

const defaultProps: AvatarFieldProps = {
  extensions: ["jpg", "jpeg", "png"],
};

const makeSut = (props?: AvatarFieldProps) => render(<AvatarField {...defaultProps} {...props} />);

describe("<AvatarField>", () => {
  it("should render component with default props", () => {
    const wrapper = makeSut();
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render component with custom extensions", () => {
    const wrapper = makeSut({ extensions: ["jpg", "jpeg"] });
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render component with custom max file size", () => {
    const wrapper = makeSut({ ...defaultProps, maxSize: 10 * 1024 * 1024 });
    expect(wrapper.container).toMatchSnapshot();
  });
});
