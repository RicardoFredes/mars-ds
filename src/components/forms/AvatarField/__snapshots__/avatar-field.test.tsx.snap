// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarField> should render component with custom extensions 1`] = `
<div>
  <div
    class="avatar-field"
  >
    <label
      class="avatar-field__content"
    >
      <input
        accept="image/*"
        class="file-input avatar-field__input"
        data-testid="file-input"
        name="file"
        title="Foto de perfil"
        type="file"
      />
      <div
        class="avatar-field__overlay"
      >
        <span
          class="icon"
          data-name="camera"
          data-testid="camera"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 8C3 8.55 3.45 9 4 9C4.55 9 5 8.55 5 8V6H7C7.55 6 8 5.55 8 5C8 4.45 7.55 4 7 4H5V2C5 1.45 4.55 1 4 1C3.45 1 3 1.45 3 2V4H1C0.45 4 0 4.45 0 5C0 5.55 0.45 6 1 6H3V8Z"
              fill="currentColor"
            />
            <circle
              cx="13"
              cy="14"
              fill="currentColor"
              r="3"
            />
            <path
              clip-rule="evenodd"
              d="M17.83 6H21C22.1 6 23 6.9 23 8V20C23 21.1 22.1 22 21 22H5C3.9 22 3 21.1 3 20V9.72C3.3 9.89 3.63 10 4 10C5.1 10 6 9.1 6 8V7H7C8.1 7 9 6.1 9 5C9 4.63 8.89 4.3 8.72 4H15.12C15.68 4 16.22 4.24 16.59 4.65L17.83 6ZM8 14C8 16.76 10.24 19 13 19C15.76 19 18 16.76 18 14C18 11.24 15.76 9 13 9C10.24 9 8 11.24 8 14Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <p
          class="text text--size-sm"
        >
          Adicionar foto
        </p>
      </div>
    </label>
    <span
      class="caption avatar-field__caption"
      data-testid="avatar-field-caption"
    >
      Arquivos *.jpg, *.jpeg
      <br />
      Tamanho máximo 1 MB
    </span>
  </div>
</div>
`;

exports[`<AvatarField> should render component with custom max file size 1`] = `
<div>
  <div
    class="avatar-field"
  >
    <label
      class="avatar-field__content"
    >
      <input
        accept="image/*"
        class="file-input avatar-field__input"
        data-testid="file-input"
        name="file"
        title="Foto de perfil"
        type="file"
      />
      <div
        class="avatar-field__overlay"
      >
        <span
          class="icon"
          data-name="camera"
          data-testid="camera"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 8C3 8.55 3.45 9 4 9C4.55 9 5 8.55 5 8V6H7C7.55 6 8 5.55 8 5C8 4.45 7.55 4 7 4H5V2C5 1.45 4.55 1 4 1C3.45 1 3 1.45 3 2V4H1C0.45 4 0 4.45 0 5C0 5.55 0.45 6 1 6H3V8Z"
              fill="currentColor"
            />
            <circle
              cx="13"
              cy="14"
              fill="currentColor"
              r="3"
            />
            <path
              clip-rule="evenodd"
              d="M17.83 6H21C22.1 6 23 6.9 23 8V20C23 21.1 22.1 22 21 22H5C3.9 22 3 21.1 3 20V9.72C3.3 9.89 3.63 10 4 10C5.1 10 6 9.1 6 8V7H7C8.1 7 9 6.1 9 5C9 4.63 8.89 4.3 8.72 4H15.12C15.68 4 16.22 4.24 16.59 4.65L17.83 6ZM8 14C8 16.76 10.24 19 13 19C15.76 19 18 16.76 18 14C18 11.24 15.76 9 13 9C10.24 9 8 11.24 8 14Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <p
          class="text text--size-sm"
        >
          Adicionar foto
        </p>
      </div>
    </label>
    <span
      class="caption avatar-field__caption"
      data-testid="avatar-field-caption"
    >
      Arquivos *.jpg, *.jpeg, *.png
      <br />
      Tamanho máximo 10 MB
    </span>
  </div>
</div>
`;

exports[`<AvatarField> should render component with default props 1`] = `
<div>
  <div
    class="avatar-field"
  >
    <label
      class="avatar-field__content"
    >
      <input
        accept="image/*"
        class="file-input avatar-field__input"
        data-testid="file-input"
        name="file"
        title="Foto de perfil"
        type="file"
      />
      <div
        class="avatar-field__overlay"
      >
        <span
          class="icon"
          data-name="camera"
          data-testid="camera"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 8C3 8.55 3.45 9 4 9C4.55 9 5 8.55 5 8V6H7C7.55 6 8 5.55 8 5C8 4.45 7.55 4 7 4H5V2C5 1.45 4.55 1 4 1C3.45 1 3 1.45 3 2V4H1C0.45 4 0 4.45 0 5C0 5.55 0.45 6 1 6H3V8Z"
              fill="currentColor"
            />
            <circle
              cx="13"
              cy="14"
              fill="currentColor"
              r="3"
            />
            <path
              clip-rule="evenodd"
              d="M17.83 6H21C22.1 6 23 6.9 23 8V20C23 21.1 22.1 22 21 22H5C3.9 22 3 21.1 3 20V9.72C3.3 9.89 3.63 10 4 10C5.1 10 6 9.1 6 8V7H7C8.1 7 9 6.1 9 5C9 4.63 8.89 4.3 8.72 4H15.12C15.68 4 16.22 4.24 16.59 4.65L17.83 6ZM8 14C8 16.76 10.24 19 13 19C15.76 19 18 16.76 18 14C18 11.24 15.76 9 13 9C10.24 9 8 11.24 8 14Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <p
          class="text text--size-sm"
        >
          Adicionar foto
        </p>
      </div>
    </label>
    <span
      class="caption avatar-field__caption"
      data-testid="avatar-field-caption"
    >
      Arquivos *.jpg, *.jpeg, *.png
      <br />
      Tamanho máximo 1 MB
    </span>
  </div>
</div>
`;
